### PR TITLE
rustbuild: fix building std for musl targets

### LINKED ISSUE #36143: rustbuild incorrectly builds std for musl targets
It copies the crt\*.o objects from the *host* system libraries when it, instead, should copy them from the musl-root. [Related code](https://github.com/rust-lang/rust/blob/71ee82a8aa0c02fc2c73e84f40bdb55512d10938/src/bootstrap/compile.rs#L103-L107)

It asks rustc to pass -ldl -lpthread and -lrt to the linker when it shouldn't because those are already included in musl libc.a. [Related code](https://github.com/rust-lang/rust/blob/71ee82a8aa0c02fc2c73e84f40bdb55512d10938/src/libstd/build.rs#L38-L42)

cc @alexcrichton

### PR DESCRIPTION
**Pull request by japaric**
closes #36143

r? @alexcrichton

### CODE REVIEW

File: src/libstd/build.rs [hunk 1]
```diff
@@ -35,7 +35,7 @@ fn main() {
             println!("cargo:rustc-link-lib=dl");
             println!("cargo:rustc-link-lib=log");
             println!("cargo:rustc-link-lib=gcc");
-        } else {
+        } else if !target.contains("musl") {
```
Comment (japaric) [inline src/libstd/build.rs]: hold on. we still need to pass these for mips-musl

File: src/bootstrap/compile.rs [hunk 2]
```diff
@@ -93,16 +93,16 @@ pub fn std_link(build: &Build,
...
-        copy_third_party_objects(build, target, &libdir);
+        copy_third_party_objects(build, &libdir);
     }
 }
 
 /// Copies the crt(1,i,n).o startup objects
 ///
 /// Only required for musl targets that statically link to libc
-fn copy_third_party_objects(build: &Build, target: &str, into: &Path) {
+fn copy_third_party_objects(build: &Build, into: &Path) {
```
Comment (alexcrichton) [inline src/bootstrap/compile.rs]: Can you change the name here to `copy_musl_third_party_object` as this only works for musl now?

**Comment by japaric** [comment]
> Can you change the name here to copy_musl_third_party_object

Done

**Comment by alexcrichton** [comment]
@bors: r+

This makes me realize we need to move `musl-root` to the target-specific configuration as well!

**Pull request merged**